lib/s3-compat.js

```diff
@@ -0,0 +1,2051 @@
+                  ' - debugging input');
+
+    // Handle x-amz-acl header
+    var s3Acl = req.headers['x-amz-acl'];
+    if (s3Acl) {
+        var mantaRoles = S3_ACL_TO_MANTA_ROLES[s3Acl.toLowerCase()];
+        if (mantaRoles !== undefined) {
+            // Mark this as an S3 ACL operation for special role handling
+            req._s3AclOperation = true;
+            req._s3AclRoles = mantaRoles;
+
+            // Also set the role-tag header for backward compatibility
+            if (mantaRoles.length > 0) {
+                req.headers['role-tag'] = mantaRoles.join(',');
+            } else {
+                // For 'private' ACL, set empty role-tag to clear roles
+                req.headers['role-tag'] = '';
+            }
+
+            req.log.debug({
+                s3Acl: s3Acl,
+                mantaRoles: mantaRoles,
+                roleTag: req.headers['role-tag'],
+                isAclOperation: true
+            }, 's3RoleTranslator: translated S3 ACL to Manta roles');
+        }
+    }
+
+    // Handle S3 grant headers (basic implementation)
+    var grantHeaders = ['x-amz-grant-read',
+                        'x-amz-grant-write',
+                        'x-amz-grant-full-control'];
+    var grantRoles = [];
+
+    grantHeaders.forEach(function (header) {
+        var grantValue = req.headers[header];
+        if (grantValue) {
+            // Parse grant value and convert to roles
+            var roles = parseS3GrantToRoles(header, grantValue);
+            grantRoles = grantRoles.concat(roles);
+        }
+    });
+
+    if (grantRoles.length > 0) {
+        var existingRoles = req.headers['role-tag'] ?
+            req.headers['role-tag'].split(',') : [];
+        var allRoles = existingRoles.concat(grantRoles);
+        // Remove duplicates
+        var uniqueRoles = allRoles.filter(function (role, index) {
+            return (allRoles.indexOf(role) === index);
+        });
+        req.headers['role-tag'] = uniqueRoles.join(',');
+        req.log.debug({
+            grantHeaders: grantHeaders,
+            grantRoles: grantRoles,
+            finalRoles: uniqueRoles
+        }, 's3RoleTranslator: translated S3 grants to Manta roles');
+    }
+
+    next();
+}
+
+/*
+ * parseS3GrantToRoles(header, value)
+ * Translates a S3 ACL grant header to the appropiate Manta role.
+ * Arguments:
+ *     header: S3 ACL grant header.
+ *     value:  String uri to represents the user/scope to apply this ACL.
+ * Returns:
+ *     Array of Manta roles associated with the grant ACL header.
+ * Notes:
+ *    Each of this roles and their associated policies must be created ahead of
+ *    time.
+ */
+function parseS3GrantToRoles(header, value) {
+    var roles = [];
+
+    // Simple implementation - can be extended for more complex grant parsing
+    switch (header) {
+        case 'x-amz-grant-read':
+            if (
+                value.includes(
+                    'uri="http://acs.amazonaws.com/groups/global/AllUsers"')) {
```

Comment:
Is it possible for both includes AllUsers *AND* AuthenticatedUsers to be inthe same value?  IF SO, granting `public-reader` is a good risk-reduction, but maybe it's going to kneecap the client who may be Authenticated?

Or are AllUsers and AuthenticatedUsers (supposed to be?) mutually exclusive in this case?